examples/vision/nerf.py

```diff
@@ -0,0 +1,810 @@
+    # r(t) = o+td -> Building the "t" here.
+    t_vals = tf.linspace(near, far, num_samples)
+    if rand:
+        # Inject uniform noise into the sample space to make the sampling
+        # continuous.
+        shape = list(ray_origins.shape[:-1]) + [num_samples]
+        noise = tf.random.uniform(shape=shape) * (far - near) / num_samples
+        t_vals = t_vals + noise
+
+    # r(t) = o + td -> Building the "r" here.
+    rays = ray_origins[..., None, :] + (
+        ray_directions[..., None, :] * t_vals[..., None]
+    )
+    rays_flat = tf.reshape(rays, [-1, 3])
+
+    # Apply positional encdoing.
+    rays_flat = position_encode(rays_flat)
+
+    return (rays_flat, t_vals)
+
+
+def map_fn(pose):
+    """
+    Maps pose to flattened rays and sample points.
+    Args:
+        - pose: The pose matrix of the camera.
+
+    Returns:
+        The flattened rays and sample points corresponding to the
+        camera pose.
+    """
+    # Get the ray origin and directions.
+    (ray_origins, ray_directions) = get_rays(height=H, width=W, focal=focal, pose=pose)
+
+    # Render the rays from the origin and direction.
+    (rays_flat, t_vals) = render_flat_rays(
+        ray_origins=ray_origins,
+        ray_directions=ray_directions,
+        near=2.0,
+        far=6.0,
+        num_samples=NUM_SAMPLES,
+        rand=True,
+    )
+    return (rays_flat, t_vals)
+
+
+# Create the training split.
+split_index = int(num_images * 0.8)
+
+# Split the images into training and validation.
+train_images = images[:split_index]
+val_images = images[split_index:]
+
+# Split the poses into training and validation.
+train_poses = poses[:split_index]
+val_poses = poses[split_index:]
+
+# Make the training pipeline.
+train_img_ds = tf.data.Dataset.from_tensor_slices(train_images)
+train_pose_ds = tf.data.Dataset.from_tensor_slices(train_poses)
+
+# Map the pose ds to get "flat_rays" and "t_vals" as dataset.
+train_ray_ds = train_pose_ds.map(map_fn, num_parallel_calls=AUTO)
+
+# Zip the image and rays dataset together.
+training_ds = tf.data.Dataset.zip((train_img_ds, train_ray_ds))
+
+# Shuffle and batch the dataset.
+train_ds = (
+    training_ds.shuffle(BATCH_SIZE)
+    .batch(BATCH_SIZE, drop_remainder=True, num_parallel_calls=AUTO)
+    .prefetch(AUTO)
+)
+
+# Make the validation pipeline.
+val_img_ds = tf.data.Dataset.from_tensor_slices(val_images)
+val_pose_ds = tf.data.Dataset.from_tensor_slices(val_poses)
+
+# Map the pose ds to get "flat_rays" and "t_vals" as dataset.
+val_ray_ds = val_pose_ds.map(map_fn, num_parallel_calls=AUTO)
+
+# Zip the image and rays dataset together.
+validation_ds = tf.data.Dataset.zip((val_img_ds, val_ray_ds))
+
+# Shuffle and batch the dataset.
+val_ds = (
+    validation_ds.shuffle(BATCH_SIZE)
+    .batch(BATCH_SIZE, drop_remainder=True, num_parallel_calls=AUTO)
+    .prefetch(AUTO)
+)
+
+"""
+## NeRF model
+
+The model is a multi-layer perceptron. It has relu as its non-linearity. 
+
+An exerpt from the paper:
```

Comment:
"excerpt". Tip: to catch typos, paste the example content into a Google doc and take a look at what gets highlighted.

Reply:
Neat trick. 😄